src/osd/PrimaryLogPG.cc

```diff
@@ -8602,6 +8609,47 @@ void PrimaryLogPG::finish_ctx(OpContext *ctx, int log_op_type, int result)
 	   << dendl;
   utime_t now = ceph_clock_now();
 
+  // Drop the reference if deduped chunk is modified
+  if (ctx->new_obs.oi.is_dirty() &&
+    (ctx->obs->oi.has_manifest() && ctx->obs->oi.manifest.is_chunked()) &&
+    // If a clone is creating, ignore dropping the reference for manifest object
+    !ctx->delta_stats.num_object_clones &&
+    ctx->new_obs.oi.size != 0 && // missing, redirect and delete
+    !ctx->cache_operation &&
+    log_op_type != pg_log_entry_t::PROMOTE) {
+    object_manifest_t updated_chunks;
```

Comment:
I don't like this.  This manifest in fact only contains chunks we plan on dropping, which is very confusing.  Moreover, all you are doing is iterating over the set of chunks and checking them against the clean_regions object.  Why not simply pass clean_regions into dec_refcount_by_dirty and then into calc_refs_to_drop_on_modify along with the clone object_manifest_t and handle it there?  You'd be able move another bit of computation into something you can actually unittest,  you avoid misleading the reader, and you reduce the amount of clutter in finish_ctx.

At the very least, call it chunks_to_remove or similar and move this whole bit of logic into dec_refcount_by_dirty.